Rust rewrite

jonner opened the pull request:
So, I think my rust rewrite branch is now basically feature complete. The final bits that I just added were the 'lsmdev' symlink and finishing the install process and rpm generation. I also made a few changes suggested by Marc-Andre, but those are mostly things to make the code more rust-like. I bumped the version to 1.0.0 to ensure that the version is higher than the last bash release. I might have used a simpler versioning scheme, but apparently cargo requires you to use a X.Y.Z versioning scheme. 

The functionality is (I think) pretty much identical to the existing implementation, and it has a few minor extras compared to the current implementation:
 - test infrastructure to prove that the results are correct
 - auto-generated bash completion scripts for both mdevctl and lsmdev

The build process is just a little bit more complex than the old version, but not too much. In essence, it uses cargo for building everything, but then uses a Makefile as an "installer". As part of the cargo build process, I generate a Makefile with the appropriate file locations for the generated files, etc. So the basic process is:
```
$ cargo build
# make install
```
or 
```
$ cargo build
$ make rpm
# dnf install $RPM
```

cohuck commented on README.md:
```diff
@@ -19,8 +19,11 @@ https://github.com/mdevctl/mdevctl
+used for installing the executable and all support files into your system.  On
+RPM based systems, you can run `make rpm` then install the resulting package.
+Otherwise, run `make install`.
 
```
I'm wondering whether we should support "make help" as well... might be overkill, though.

jonner replied:
What would `make help` do in this case?

cohuck replied:
Just something simple, like

```
Supported make targets:
  rpm - create an .rpm package
  install - install the program and auxiliary files
```
As said, that might well be overkill.

cohuck commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "mdevctl"
-version = "0.1.0"
+version = "1.0.0"
```
We probably should make clear from the start that this is not semver, and the numbers do not have any particular meaning; too bad that we are stuck with the x.y.z convention.
Maybe do a monotonic increase of the second or the third digit every time we do a release and bump the first/second digit once we would reach 0xff?

jonner replied:
Yeah, I'm open to versioning suggestions. I was thinking about just bumping the 'y' number on each release for now, but other approaches are fine as well. I'm not sure whether it's necessary to be explicit that this is not a semantic version. Nothing is really consuming 'mdevctl' in a manner that the version number will really make a difference.

cohuck replied:
I'd probably leave a note to what we do (your scheme is fine), just to avoid any possible confusion.

cohuck commented:
On the whole, this looks good to me; would be great to get a second opinion, though, especially from someone with Rust experience (I don't have any real practical experience.)

jonner commented on build.rs:
```diff
@@ -1,18 +1,53 @@
+    let mdevctl_bash_path = outdir.join("mdevctl.bash");
+
+    let mut lsmdev_cli = cli::LsmdevOptions::clap();
+    lsmdev_cli.gen_completions("lsmdev", Shell::Bash, &outdir);
```
Same here.

awilliam commented on Makefile.in:
```diff
@@ -0,0 +1,49 @@
+DATADIR=$(PREFIX)/share
+MANDIR=$(DATADIR)/man
+BASH_COMPLETION_DIR=$(DATADIR)/bash-completion/completions
+SPECFILE=@@spec_dir@@/rust-mdevctl.spec
```
I can't make this work, make rpm or srpm fails with:
```
$ make srpm
cargo build
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
# since we generate a Makefile, we have to use --no-verify to stop cargo
# from complaining that we modified the source directory during build.
cargo package --allow-dirty --no-verify
   Packaging mdevctl v1.0.0 (/home/alwillia/Work/mdevctl.git)
    Updating crates.io index
rpmbuild -bs --define "_sourcedir /home/alwillia/Work/mdevctl.git/target/package" \
	--define "_specdir /home/alwillia/Work/mdevctl.git" \
	--define "_builddir /home/alwillia/Work/mdevctl.git" \
	--define "_srcrpmdir /home/alwillia/Work/mdevctl.git" \
	--define "_rpmdir /home/alwillia/Work/mdevctl.git" @@spec_dir@@/rust-mdevctl.spec
error: failed to stat /home/alwillia/Work/mdevctl.git/@@spec_dir@@/rust-mdevctl.spec: No such file or directory
make: *** [Makefile:25: srpm] Error 1
```

jonner replied:
weird. it was working before. I'll investigate.

awilliam replied:
Gets further now, but I still get crate dependencies that are not resolved running builddep on the srpm:
```
$ sudo dnf builddep /home/alwillia/Work/mdevctl.git/rust-mdevctl-1.0.0-1.fc33.src.rpm
Last metadata expiration check: 1:31:39 ago on Thu 27 May 2021 09:37:51 AM MDT.
Package rust-packaging-16-1.fc33.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
$ make rpm
cargo package --allow-dirty --no-verify
   Packaging mdevctl v1.0.0 (/home/alwillia/Work/mdevctl.git)
    Updating crates.io index
rpmbuild -bb --define "_sourcedir /home/alwillia/Work/mdevctl.git/target/package" \
	--define "_specdir /home/alwillia/Work/mdevctl.git" \
	--define "_builddir /home/alwillia/Work/mdevctl.git" \
	--define "_srcrpmdir /home/alwillia/Work/mdevctl.git" \
	--define "_rpmdir /home/alwillia/Work/mdevctl.git" rust-mdevctl.spec
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.QnzGUY
+ umask 022
+ cd /home/alwillia/Work/mdevctl.git
+ cd /home/alwillia/Work/mdevctl.git
+ rm -rf mdevctl-1.0.0
+ /usr/bin/gzip -dc /home/alwillia/Work/mdevctl.git/target/package/mdevctl-1.0.0.crate
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd mdevctl-1.0.0
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ set -eu
+ /usr/bin/mkdir -p .cargo
+ cat
+ /usr/bin/rm -f Cargo.lock
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%generate_buildrequires): /bin/sh -e /var/tmp/rpm-tmp.9TomxV
+ umask 022
+ cd /home/alwillia/Work/mdevctl.git
+ cd mdevctl-1.0.0
+ /usr/bin/cargo-inspector -BR Cargo.toml
+ /usr/bin/cargo-inspector -TR Cargo.toml
+ RPM_EC=0
++ jobs -p
+ exit 0
error: Failed build dependencies:
	(crate(anyhow/default) >= 1.0.0 with crate(anyhow/default) < 2.0.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(env_logger/default) >= 0.8.3 with crate(env_logger/default) < 0.9.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(log/default) >= 0.4.0 with crate(log/default) < 0.5.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(serde_json/default) >= 1.0.0 with crate(serde_json/default) < 2.0.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(serde_json/preserve_order) >= 1.0.0 with crate(serde_json/preserve_order) < 2.0.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(structopt/default) >= 0.3.13 with crate(structopt/default) < 0.4.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(tempdir/default) >= 0.3.0 with crate(tempdir/default) < 0.4.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(uuid/default) >= 0.8.2 with crate(uuid/default) < 0.9.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
	(crate(uuid/v4) >= 0.8.2 with crate(uuid/v4) < 0.9.0) is needed by rust-mdevctl-1.0.0-1.fc33.x86_64
Wrote: /home/alwillia/Work/mdevctl.git/rust-mdevctl-1.0.0-1.fc33.buildreqs.nosrc.rpm
make: *** [Makefile:31: rpm] Error 11
```

connorkuehl replied:
Crates are independently packaged in Fedora starting only with Fedora 34 (since I see the f33 references in your output), so the dependencies will be unavailable unless you're using Fedora 34 or newer.

awilliam commented on Makefile.in:
```diff
@@ -0,0 +1,49 @@
+all:
+	@echo "Run 'make install' to install the program"
+
+$(SPECFILE): rust-mdevctl.spec.in
```
I think we need a more user friend specfile target, generally when uploading to Fedora I 'make mdevctl.spec' then use spectool to fetch the upstream source tarball, and rpmbuild from the spec file.  Digging the generated spec file out of a generated directory is not conducive to this.
```
$ find ./ -name rust-mdevctl.spec
./target/debug/build/mdevctl-7acebde1237c5868/out/rust-mdevctl.spec
```
Are we also changing the name to rust-mdevctl?

jonner replied:
Basically I was following the fedora rust packaging guidelines here: https://docs.fedoraproject.org/en-US/packaging-guidelines/Rust/

It seems that rust applications have a srpm named rust-$PACKAGE, and they generate binary rpms named $PACKAGE, $PACKAGE-debuginfo and rust-$PACKAGE-debugsource. I'm still figuring out the rust packaging process.

As for the generated file location: I could dump it in the source dir, but cargo gets unhappy if you change the source dir as part of the build. I'm already violating that by generating a Makefile, so I guess it wouldn't hurt to dump the spec file here well.

awilliam commented on rust-mdevctl.spec.in:
```diff
@@ -0,0 +1,64 @@
+# Upstream license specification: LGPL-2.1
+License:        LGPLv2
+URL:            https://crates.io/crates/mdevctl
+Source:         %{crates_source}
```
We need for spectool to be able to download sources, so we need real urls for these.  I don't see that mdevctl exists on crates.io (do all published Rust crates live there?), so this currently results in:
```
$ spectool -g -R ./target/debug/build/mdevctl-7acebde1237c5868/out/rust-mdevctl.spec
Downloading: https://crates.io/api/v1/crates/mdevctl/1.0.0/download#/mdevctl-1.0.0.crate
Download failed:
404 Client Error: Not Found for url: https://crates.io/api/v1/crates/mdevctl/1.0.0/download#/mdevctl-1.0.0.crate
-   0.0 B Elapsed Time: 0:00:00              
```
Is rpmbuild going to need a tarball of this or is a crate the source format?

jonner replied:
The .crate file is the source format (the fedora rust-packaging macros expect this format). As far as I know, it's basically a source tarball.

It seems to be more-or-less standard for rust projects to publish their releases on crates.io, but I haven't done so (yet?) partly because we haven't had an official release of the rust code yet, and partly because I didn't know whether publishing a commandline utility (as opposed to e.g. a rust library) on crates.io is useful since nobody will re-use it in their rust application. But we certainly could publish it there.

awilliam commented on rust-mdevctl.spec.in:
```diff
@@ -0,0 +1,64 @@
+%endif
+
+%changelog
+* Tue Apr 27 10:55:11 CDT 2021 Jonathon Jongsma <jjongsma@redhat.com> - 0.1.0-1
```
I wonder if one of the package validation tools is going to choke that this version doesn't match the 1.0.0 version that I think we're expecting to be generated.

awilliam commented on Makefile.in:
```diff
@@ -0,0 +1,49 @@
+SPECFILE=@@spec_dir@@/rust-mdevctl.spec
+
+archive:
+	# since we generate a Makefile, we have to use --no-verify to stop cargo
```
These need to be un-indented or else they show up in the build log.

cohuck commented:
@collinwalling @zhuoyingcai do you think this will work for your callouts support?

collinwalling commented:
I completely missed this! I should've paid more attention to the PR tab. I'll take a look at this today, though from what I saw from the code in GitLab, things looked pretty sane... I'm not sure what has changed between then-and-now, and I've yet to take a look at the code after getting a better understanding of the language.

jonner commented:
I just force-pushed my branch with changes to address many of the comments above.  Hopefully that didn't cause any problems for anyone. @awilliam the rpm generation stuff should work now.

connorkuehl commented on src/mdev.rs:
```diff
@@ -0,0 +1,472 @@
+    }
+
+    // get parent and propagate a consistent error to the caller if absent
+    pub fn parent(&self) -> Result<&String> {
```
Could this return `&str` instead? Is there a specific reason you're returning a borrowed owned type (String) vs just a borrowed type (&str)?

jonner replied:
Something like this? 

```suggestion
    pub fn parent(&self) -> Result<&str> {
                self.parent.map(|x| x.as_str()).ok_or_else(|| {
```
(replaced as_ref() with map())

It's not clear to me what we would gain from a change like this.

connorkuehl commented on src/mdev.rs:
```diff
@@ -0,0 +1,472 @@
+    }
+
+    // get mdev_type and propagate a consistent error to the caller if absent
+    pub fn mdev_type(&self) -> Result<&String> {
```
Same comment as `pub fn parent`.

connorkuehl commented on src/main.rs:
```diff
@@ -0,0 +1,648 @@
+}
+
+/// parse command line arguments and dispatch to command-specific functions
+fn main() -> Result<()> {
```
Instead of letting the Result bubble all the way back down past the main method, I would recommend matching in the main method to catch the error and print it out, and then if it has a cause (`cause()` method) then recursively print that out.

Otherwise some day there may be complaints from users about "ugly Rust stack traces" :sweat_smile:

jonner replied:
So, when I run the executable with an error case, it just prints out the error message as expected. Are you sure that this will print out rust stack traces? I've never had it print a stack trace for error results...

connorkuehl replied:
Oops, no, I don't think it will do a Rust panic backtrace, it will just print the bubbled down error.

If you have errors that encapsulate other errors (such as wrapping an io::Error with a contextual error, with information like line number or something), this can be a good way of preserving the "error trace" to track down exactly where the "point of first bubbling down" was.

Sometimes with bubbling errors out with a sufficiently deep call stack, it can be trickier to determine which part of the code failed without that kind of context.

Just a user experience suggestion.

connorkuehl reviewed (commented):
@jonner I think this is some great work! I mainly reviewed this from a Rust lens and not a correctness lens for mdevctl since that's where my expertise falls off.

I think the broad strokes of my observations would be:

- Try to impl the required traits so that your structs can be serialized/deserialized with serde without any manual work. I see there's some of that here. That's not to say that yours doesn't work, of course not, but serde is quite mature and in this way you might be able to benefit from them with less work on your part.
- Consider returning an idiomatically borrowed type like &str over &String if possible
- Consider catching errors in main and printing out the error chain for the user to avoid exposing them to a Rust backtrace
- There are a number of unwraps and expects in methods that return Results that could be easily replaced with anyhow errors to avoid any unsightly stack traces as alluded to in the above suggestion.
- Some of your functions take a lot of arguments. This isn't an issue in and of itself, but I would suggest encoding them in structs to model these arguments and see how you feel about that. Not a deal breaker for anyone; clippy is just a big complainer :-)

Overall, nice job :+1: All of my comments are very minor and I didn't see any real red flags.

jonner commented on src/main.rs:
```diff
@@ -0,0 +1,648 @@
+mod tests;
+
+/// Format a map of mediated devices into a json string
+fn format_json(devices: HashMap<String, Vec<MDev>>) -> Result<String> {
```
So, I really wanted to do this, but I'm not sure that it's possible with the JSON format that mdevctl currently uses. The main problem I ran into is that it uses a lot of variable json field names. For example, this is the JSON representation of a mediated device: 

```
{
  "b845d580-7a6b-453d-b22d-078d5be12ccd": {
    "mdev_type": "i915-GVTg_V5_8",
    "start": "auto"
  }
}
````

Notice how the UUID of the device is not the value of a property of the object, but is rather the name of key to an object containing the rest of the properties of the device. If instead the JSON representation was something like this, it would be a lot easier to implement automatic serialize/deserialize for a struct:

````
{
  "uuid": "b845d580-7a6b-453d-b22d-078d5be12ccd",
  "mdev_type": "i915-GVTg_V5_8",
  "start": "auto"
}
````

Similarly here's the JSON that is generated to represent a list of devices:

````
[
  {
    "0000:00:02.0": [
      {
        "b845d580-7a6b-453d-b22d-078d5be12ccd": {
          "mdev_type": "i915-GVTg_V5_8",
          "start": "auto"
        }
      },
      {
        "88a6b868-46bd-4015-8e5b-26107f82da38": {
          "mdev_type": "i915-GVTg_V5_8",
          "start": "manual"
        }
      }
    ]
  }
]
````

Note how the key to the array of devices is the name of the parent device, which is essentially an arbitrary string (well, it's the PCI address for PCI parent devices, but it's variable).

Maybe there's a way to get automatic serialization/deserialization with this format, but I couldn't figure it out after a quick glance. So since I was still pretty new to rust, I went for the more straightforward method of writing manual serialization functions to get something working. Perhaps I can investigate further though. If you have ideas, I'd love to hear them.

If I was starting from scratch I would probably use a different JSON format, but I wanted to keep compatibility with the existing code.

connorkuehl replied:
Makes sense! I understand now.

jonner commented on build.rs:
```diff
@@ -0,0 +1,67 @@
+
+fn apply_template(template: &PathBuf) -> String {
+    let outdir =
+        PathBuf::from(env::var_os("OUT_DIR").expect("OUT_DIR environment variable not defined"));
```
perhaps OUT_DIR is not set while compiling the build script, because I get the following error

```
error: environment variable `OUT_DIR` not defined
  --> build.rs:10:18
   |
10 |     let outdir = env!("OUT_DIR");
   |                  ^^^^^^^^^^^^^^^
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

error: could not compile `mdevctl`

To learn more, run the command again with --verbose.

```

jonner commented on build.rs:
```diff
@@ -0,0 +1,67 @@
+        )
+}
+
+fn main() {
```
Thanks. I think it's a much better solution than I was fearing. But I can't help but feel that there must be something just a little bit better ;)

jonner commented:
> I completely missed this! I should've paid more attention to the PR tab. I'll take a look at this today, though from what I saw from the code in GitLab, things looked pretty sane... I'm not sure what has changed between then-and-now, and I've yet to take a look at the code after getting a better understanding of the language.

@collinwalling Have you had a chance to take a look yet?

zhuoyingcai commented:
I have been working on rewriting the callout function in Rust. The rewrite is in good progress so far. But we need to figure out a good location to invoke the callout function in Rust version.

When I was testing the callout function, I found that the rust version has a different output when defining a device without attributes.
`cargo run define -p matrix -t vfio_ap-passthrough`
Rust output:
```
{
  "mdev_type": "vfio_ap-passthrough",
  "start": "manual"
}
```
`mdevctl define -p matrix -t vfio_ap-passthrough`
Bash output:
```
{
  "mdev_type": "vfio_ap-passthrough",
  "start": "manual",
  "attrs": []
}
```

The callout script is giving this error on rust version:
`ap_check: Failed to read device config.`

The callout script will exit on error if it doesn't see any attributes. We would like to know which output format will be the final one, so we can make changes accordingly. Thanks!

jonner commented:
@zhuoyingcai Hmm, interesting. By definition, the original mdevctl bash implementation is probably "correct", since I'm simply trying to make my version identical to it. However, it's interesting to note that when you run `mdevctl list -u $UUID --defined --dumpjson` on that device that you just defined, it *doesn't* actually include the attrs field in _that_ output:

````
$ mdevctl list -u 06a068fc-d16f-4cb3-9d18-74e2681af094 --defined --dumpjson
{
  "mdev_type": "vfio_ap-passthrough",
  "start": "manual"
}
````

So the original bash implementation is a bit inconsistent on that matter. I can certainly change the rust implementation to include the `attrs` field in the config file for a defined device, but it seems to me that it would also be good for the callout script to be flexible enough to handle both an empty `attrs` list and a missing `attrs` list. Unless there's a reason that it's important for you to treat an empty list different than a missing list?

cohuck commented:
> So the original bash implementation is a bit inconsistent on that matter. I can certainly change the rust implementation to include the `attrs` field in the config file for a defined device, but it seems to me that it would also be good for the callout script to be flexible enough to handle both an empty `attrs` list and a missing `attrs` list. Unless there's a reason that it's important for you to treat an empty list different than a missing list?

The Rust rewrite could actually be a good opportunity to get rid of some of the existing inconsistencies... providing an empty list looks like a good idea. Having robust scripts is not a bad idea either, though :)

zhuoyingcai commented:
> @zhuoyingcai Hmm, interesting. By definition, the original mdevctl bash implementation is probably "correct", since I'm simply trying to make my version identical to it. However, it's interesting to note that when you run `mdevctl list -u $UUID --defined --dumpjson` on that device that you just defined, it _doesn't_ actually include the attrs field in _that_ output:
> 
> ```
> $ mdevctl list -u 06a068fc-d16f-4cb3-9d18-74e2681af094 --defined --dumpjson
> {
>   "mdev_type": "vfio_ap-passthrough",
>   "start": "manual"
> }
> ```
> 
> So the original bash implementation is a bit inconsistent on that matter. I can certainly change the rust implementation to include the `attrs` field in the config file for a defined device, but it seems to me that it would also be good for the callout script to be flexible enough to handle both an empty `attrs` list and a missing `attrs` list. Unless there's a reason that it's important for you to treat an empty list different than a missing list?

Reply from Matt: 
"There is no reason the script can't handle a missing attrs array (I was planning to add it anyway as yes, it should be able to tolerate both) but rather we just wanted to point out the inconsistency between the versions."

jonner commented:
@zhuoyingcai OK, thanks. I'll make the change to include the empty attrs field in my version as @cohuck suggests. Consistency is good.

jonner commented:
I apologize for the churn here. I discovered a pretty significant compatibility error in my implementation of `mdevctl list` so I fixed it. Since the branch is still not merged, I chose to rebase the implementation rather than adding a fixup commit at the end of the branch. So unfortunately it looks like a lot of stuff changed.

In summary, the original mdevctl implementation uses a slightly odd JSON format for listing devices from multiple parent devices. The root of the JSON output is an array, but that array only ever has a single element, which is an object with a separate field for each parent device. When I reimplemented it in rust, I made a bad assumption that the array would contain multiple objects -- one for each parent. But I apparently never validated my assumption by testing. 

Incorrect output:
````
[
  {
    "parent1": [ list of mdevs ]
  },
  {
    "parent2": [ list of mdevs ]
  }
]
````
Correct output:
````
[
  {
    "parent1": [ list of mdevs ],
    "parent2": [ list of mdevs ]
  }
]
````


I've now fixed it to match the old implementation and added testing for the `mdevctl list` output.

smitterl commented on src/environment.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    fn parent_base(&self) -> PathBuf {
+        self.root().join("sys/class/mdev_bus")
```
This path doesn't need to exist. As a consequence I get an error when running
```bash
# mdevctl types
Error: No such file or directory (os error 2)
```
The bash script returns empty with exit code 0.

jonner replied:
I have a local patch that fixes this. will push new changes soon.

smitterl commented on Makefile.in:
```diff
@@ -0,0 +1,48 @@
+		--define "_srcrpmdir $(PWD)" \
+		--define "_rpmdir $(PWD)" $<
+
+rpm: rust-mdevctl.spec archive
```
This command fails for me on RHEL, I don't know how to fix it.
```bash
# make rpm
cargo package --allow-dirty --no-verify
   Packaging mdevctl v1.0.0 (/root/mdevctl)
    Updating crates.io index
rpmbuild -bb --define "_sourcedir /root/mdevctl/target/package" \
	--define "_specdir /root/mdevctl" \
	--define "_builddir /root/mdevctl" \
	--define "_srcrpmdir /root/mdevctl" \
	--define "_rpmdir /root/mdevctl" rust-mdevctl.spec
error: Bad source: /root/mdevctl/target/package/%{crates_source}: No such file or directory
make: *** [Makefile:31: rpm] Error 1
```

jonner replied:
Yeah, at the moment, the rpm build only works on fedora because it requires some rust rpm packaging macros.

awilliam replied:
Do we consider this a blocker?  AIUI, it's not just fedora, but exclusively f34+ that works.  If mdevctl were EOL on all RPM distros not released in the past 6mo, that seems troublesome.

jonner replied:
Yes, it's true that the upstream rpm spec file requires f34+. That's because the upstream rpm uses the new standard [Fedora rust packaging guidelines](https://docs.fedoraproject.org/en-US/packaging-guidelines/Rust/). As far as I know, nothing precludes it from being packaged on different/older distributions, it would just require more work because it would require more ad-hoc/manual approach. I didn't feel that was important to support that in the upstream repository. Do you consider it a blocker?

awilliam replied:
Are we going to abandon mdevctl for f33 and RHEL8 (I don't think we can) or are you just imagining that those are downstream issues and they'll (you'll) create downstream specific spec files for them?  IMO the best solution would be pre-f34 compatibility upstream that could be removed over time.

jonner replied:
I guess I was assuming that the rust version of mdevctl would only be packaged for newer distributions. Distributions like f33/rhel8 that don't have the dependencies packaged yet could just stay on the old release of mdevctl (with downstream patches if necessary).

jonner replied:
I think I have a spec file that can work on rhel8/f33 now, thanks to a bit of help from Connor.

smitterl commented on Makefile.in:
```diff
@@ -0,0 +1,48 @@
+		--define "_builddir $(PWD)" \
+		--define "_srcrpmdir $(PWD)" \
+		--define "_rpmdir $(PWD)" $<
+
```
I'm always thankful for a `clean` command to help restore my test environment. :)

smitterl commented on src/tests.rs:
```diff
@@ -0,0 +1,1057 @@
+}
+
+#[test]
+fn test_list() {
```
This test fails for me on s390x:
```bash
thread 'tests::test_list' panicked at 'assertion failed: `(left == right)`
  left: `"9f579710-6ffc-4201-987a-4ffa0fb1f3a5 0000:00:02.0 vfio_ap-passthrough manual\n4a0a190f-dcf3-4def-9342-c48768f0c940 0000:00:02.0 i915-GVTg_V5_8 auto\n9f579710-6ffc-4201-987a-4ffa0fb1f3a5 0000:00:03.0 vfio_ap-passthrough manual\n"`,
 right: `"4a0a190f-dcf3-4def-9342-c48768f0c940 0000:00:02.0 i915-GVTg_V5_8 auto\n9f579710-6ffc-4201-987a-4ffa0fb1f3a5 0000:00:02.0 vfio_ap-passthrough manual\n9f579710-6ffc-4201-987a-4ffa0fb1f3a5 0000:00:03.0 vfio_ap-passthrough manual\n"`', src/tests.rs:143:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

jonner replied:
That's interesting. That implies that the BTreeMap has different ordering on s390 than on x86_64? Does it fail every time?

collinwalling replied:
I ran a `cargo test list` on my s390 machine (z13 generation, if that matters) and it passed.

Looking at the right-value from above, I don't see anything labeled "i915-GVTg_V5_8" in the `test_list` function (in fact, any i915-GVTg-V5_* is listed only under the define tests... and I didn't successfully grep for a *_8 in the tests.rs file). Is this a particular test case that I'm missing on my end? 

... or perhaps I don't understand how the tests are generated (they look static to me). I just want to see if I can replicate this on my end.

jonner replied:
@collinwalling: any function that's annotated with `#[test]` above it is treated as a test function and is run automatically when you execute `cargo test`. Each test function has a test environment object which contains a data path and a temporary dir that acts as a mock filesystem root for the test execution. The data path points to a directory that contains both input files (used for constructing the mock filesystem) and files that contain the expected output for commands. In this case, the "i915-GVTg_V5_8" string is located in tests/list/default/device2.json that is used to define a device that is injected into the mock filesystem like this:
```
    test.populate_defined_device(UUID[2], PARENT[0], "device2.json");
```

[By the way, I didn't realize that `cargo test $string` would match substrings of the test function names and therefore will run the test_list() test. For instance, `cargo test define` will run both test_define() and test_undefine()...]

smitterl replied:
I can reproduce the issue reliably. But it depends on my environment:

1. z13 LPAR
```bash
# rpm -q rust
rust-1.52.1-1.module+el8.4.0+11282+0729bac9.s390x
# cargo test 2>/dev/null |grep test_list|head -1
test tests::test_list ... FAILED
```
2. z15 LPAR
```bash
# rpm -q rust
rust-1.52.1-3.el9.s390x
# cargo test 2>/dev/null |grep test_list|head -1
test tests::test_list ... ok
```

jonner replied:
interestingly, I just managed to reproduce this myself on an x86_64 rhel8 guest. 
```
$ rpm -q rust
rust-1.49.0-1.module+el8.4.0+9446+1a463e08.x86_64
```
I'll look into it, but the only impact is that mdevctl might list devices in a slightly different order.

jonner replied:
It turns out it wasn't related to BTreeMap. Fixed locally.

smitterl commented on src/cli.rs:
```diff
@@ -0,0 +1,209 @@
+                attributes are provided via the file.\n\n\
+                Running devices are unaffected by this command."
+    )]
+    Define {
```
Help text and errors for `define` are wrong:
```bash
# mdevctl define
error: The following required arguments were not provided:
    --parent <parent>
    --uuid <uuid>

USAGE:
    mdevctl define [OPTIONS] --parent <parent> --uuid <uuid>

# mdevctl define --parent 0.0.26aa --type vfio_ccw-io
8847e3d7-10f6-42d1-baab-1301bfaafad2

# mdevctl list -d
8847e3d7-10f6-42d1-baab-1301bfaafad2 0.0.26aa vfio_ccw-io manual
```

jonner replied:
This help text is auto-generated from the structopt option parsing framework. `parent` is defined as "required unless uuid is specified", which apparently results in this slightly confusing help text.  I'm not sure whether this can be improved much with the framework I'm currently using...

smitterl commented on src/cli.rs:
```diff
@@ -0,0 +1,209 @@
+                include attributes for the device(s)."
+    )]
+    List(LsmdevOptions),
+    #[structopt(
```
Using the same uuid with different parents leads to several issues both when `-p` is given or not:

Without `-p`
```bash
[root@kernelqe3 ~]# mdevctl list -d
[2021-06-16T07:43:17Z WARN  mdevctl::mdev] Overwriting parent for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: matrix => 0.0.26aa
[2021-06-16T07:43:17Z WARN  mdevctl::mdev] Overwriting mdev type for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: vfio_ap-passthrough => vfio_ccw-io
8847e3d7-10f6-42d1-baab-1301bfaafad2 0.0.26aa vfio_ccw-io auto (active)
8847e3d7-10f6-42d1-baab-1301bfaafad2 0.0.26aa vfio_ccw-io manual (active)

# mdevctl list
8847e3d7-10f6-42d1-baab-1301bfaafad2 0.0.26aa vfio_ccw-io manual
^^ the `(defined)` suffix is missing and the device is displayed as manual start but I had defined that one as auto-start (first define device under parent p1 with auto start then define device under parent p2 with same uuid and manual start)


[root@kernelqe3 ~]# mdevctl list -d -u 8847e3d7-10f6-42d1-baab-1301bfaafad2
[2021-06-16T07:43:46Z WARN  mdevctl::mdev] Overwriting parent for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: matrix => 0.0.26aa
[2021-06-16T07:43:46Z WARN  mdevctl::mdev] Overwriting mdev type for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: vfio_ap-passthrough => vfio_ccw-io
8847e3d7-10f6-42d1-baab-1301bfaafad2 0.0.26aa vfio_ccw-io auto (active)
8847e3d7-10f6-42d1-baab-1301bfaafad2 0.0.26aa vfio_ccw-io manual (active)
```
```bash
[root@kernelqe3 ~]# mdevctl undefine -u 8847e3d7-10f6-42d1-baab-1301bfaafad2
[2021-06-16T07:46:04Z WARN  mdevctl::mdev] Overwriting parent for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: matrix => 0.0.26aa
[2021-06-16T07:46:04Z WARN  mdevctl::mdev] Overwriting mdev type for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: vfio_ap-passthrough => vfio_ccw-io
Error: Failed to remove file "/etc/mdevctl.d/0.0.26aa/8847e3d7-10f6-42d1-baab-1301bfaafad2"

Caused by:
    No such file or directory (os error 2)

[root@kernelqe3 ~]# mdevctl start  --uuid 8847e3d7-10f6-42d1-baab-1301bfaafad2
[2021-06-16T07:47:17Z WARN  mdevctl::mdev] Overwriting parent for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: matrix => 0.0.26aa
[2021-06-16T07:47:17Z WARN  mdevctl::mdev] Overwriting mdev type for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: vfio_ap-passthrough => vfio_ccw-io
Error: Device already exists
```
With `-p`
```bash
# mdevctl list -d -p matrix
[2021-06-16T07:49:20Z WARN  mdevctl::mdev] Overwriting parent for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: matrix => 0.0.26aa
[2021-06-16T07:49:20Z WARN  mdevctl::mdev] Overwriting mdev type for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: vfio_ap-passthrough => vfio_ccw-io
```
```bash
# mdevctl start -p matrix -u 8847e3d7-10f6-42d1-baab-1301bfaafad2
[2021-06-16T07:52:57Z WARN  mdevctl::mdev] Overwriting parent for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: matrix => 0.0.26aa
[2021-06-16T07:52:57Z WARN  mdevctl::mdev] Overwriting mdev type for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: vfio_ap-passthrough => vfio_ccw-io
Error: Device already exists
```
```bash
# mdevctl undefine -p matrix -u 8847e3d7-10f6-42d1-baab-1301bfaafad2
[2021-06-16T07:53:33Z WARN  mdevctl::mdev] Overwriting parent for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: matrix => 0.0.26aa
[2021-06-16T07:53:33Z WARN  mdevctl::mdev] Overwriting mdev type for mdev 8847e3d7-10f6-42d1-baab-1301bfaafad2: vfio_ap-passthrough => vfio_ccw-io
Error: Failed to remove file "/etc/mdevctl.d/0.0.26aa/8847e3d7-10f6-42d1-baab-1301bfaafad2"

Caused by:
    No such file or directory (os error 2)
```

cohuck replied:
Quick question: did that work with the bash version?

smitterl replied:
The output format for list of running devices is different, the rust implementation displays the start type:
1. Bash script:
```bash
# mdevctl list
e64841ab-e98c-45f6-952a-a06b6da0ea2a 0.0.26ab vfio_ccw-io (defined)
```
2. Rust:
```bash
# mdevctl list
f5bcd615-f1df-445e-9108-a344889eb4c8 0.0.26aa vfio_ccw-io manual (defined)
```

cohuck replied:
I'd argue that the new format is actually an improvement. Does anyone actively parse that output? I would expect any tools on top to work with the json format.

jonner replied:
Hopefully you're only referring to start type and not the uuid/parent differences? Did you execute these two commands when different mdevs were active?

smitterl replied:
LOL I mean the start type, yes, sorry that I wasn't clear here - and I agree with Cornelia that it's an improvement.

smitterl replied:
> Quick question: did that work with the bash version?

Yes

jonner replied:
fixed locally

smitterl commented on src/cli.rs:
```diff
@@ -0,0 +1,209 @@
+        structopt::clap::AppSettings::UnifiedHelpMessage,
+    ]
+)]
+pub enum MdevctlCommands {
```
For the bash script version is a command:
```bash
# mdevctl version
0.78
```
while in Rust it has become an option:
```bash
# mdevctl --version
mdevctl 1.0.0
```

jonner replied:
Indeed. This is a side-effect of the option parsing framework that I'm using. Is this important? I can try to add a subcommand for version.

jonner replied:
I'm inclined to just leave this as an option rather than trying to add a subcommand `version`. I think it's unlikely that anybody requires absolute stability in the CLI interface for printing the mdevctl version. The fact that it has changed from `mdevctl version` to `mdevctl --version` is a small change, and it's easily discoverable from the help text. So unless anybody objects, I'd rather stay with the auto-generated `--version` approach.

smitterl replied:
I agree that it doesn't need to be equivalent with bash script because I can't think of a scenario where this would be important. I still hope you agree that it's good to mention all differences between the two implementations.

smitterl commented on build.rs:
```diff
@@ -0,0 +1,67 @@
+        PathBuf::from(env::var_os("OUT_DIR").expect("OUT_DIR environment variable not defined"));
+
+    // generate bash completions for both executables
+    cli::MdevctlCommands::clap().gen_completions("mdevctl", Shell::Bash, &outdir);
```
Completions leak internal command `start-parent-mdevs`:
```bash
# mdevctl st
start               start-parent-mdevs  stop
```

jonner replied:
as far as I can tell, the framework doesn't allow us to avoid generating a completion for the hidden subcommand. I don't think that this is particularly important, but I've opened a request with the framework I'm using: https://github.com/clap-rs/clap/issues/2541

smitterl commented on src/cli.rs:
```diff
@@ -0,0 +1,209 @@
+                or manual. \n\n\
+                Running devices are unaffected by this command."
+    )]
+    Modify {
```
Usage string is incorrect, it actually differs between the error message and the help message:
```bash
# mdevctl modify -u fcbfa65a-7459-44b3-93c5-d47bbd48b5d0 --addattr assign_adapter 0x01
error: Found argument '0x01' which wasn't expected, or isn't valid in this context

USAGE:
    mdevctl modify --addattr <addattr> --uuid <uuid>

# mdevctl modify -h
[...]
USAGE:
    mdevctl modify [OPTIONS] --uuid <uuid>
[...]
```

jonner replied:
The error message is correct but perhaps not very helpful. If you specify `--addattr`, you must also specify the attribute value with `--value`. The usage strings are autogenerated, but they differ because the first one is providing context to the specific command you were trying to execute (which includes the `--addattr`) option, whereas the second one provides general usage where you may not want to use `--addattr`.

I just whipped up a patch that tries to clarify things slightly. Does this help?

```
mdevctl modify --help
[...]
USAGE:
    mdevctl modify [OPTIONS] --uuid <uuid>

OPTIONS:
        --addattr <attr_name>    add a new attribute
[...]
        --value <attr_value>     Value for the attribute specified by --addattr
```

Notice that I customized the display name of the value so that instead of e.g. `--addattr <addattr>` it now says `--addattr <attr_name>`.  I also added a requirement that if you specify `--addattr`, you must also specify `--value`. So your previous command now results in the following output:
```
$ mdevctl modify -u fcbfa65a-7459-44b3-93c5-d47bbd48b5d0 --addattr assign_adapter 0x01
error: Found argument '0x01' which wasn't expected, or isn't valid in this context

USAGE:
    mdevctl modify --addattr <attr_name> --uuid <uuid> --value <attr_value>

For more information try --help
```

smitterl commented on src/main.rs:
```diff
@@ -0,0 +1,666 @@
+}
+
+/// convert 'list' command arguments into a text output
+fn list_command_helper(
```
The json dump for a single device is different. Bash script returns only the information below the uuid, whereas today's rust implementation returns a list and the entry starts at the parent. The bash script output allows to dump definition to a file and reuse it as-is for definition, the rust implementation doesn't. I think this is a blocker.
(Please ignore any exact ids in output below.)
```bash
# mdevctl --version
mdevctl 1.0.0
# mdevctl list -du 3b8bd028-2e86-4362-8b14-b3fd722b23fc --dumpjson
[
  {
    "0.0.26aa": [
      {
        "3b8bd028-2e86-4362-8b14-b3fd722b23fc": {
          "mdev_type": "vfio_ccw-io",
          "start": "manual",
          "attrs": []
        }
      }
    ]
  }
]
```
```bash
{
  "mdev_type": "vfio_ccw-io",
  "start": "manual"
}
```

For empty attributes list, in the rust there's an additional `attrs` but that's okay I think because it is still a valid definition for 0.78, I could successfully define a device though the empty list was defined.

jonner replied:
oh, interesting. Good catch.

jonner replied:
fixed locally

smitterl commented on src/main.rs:
```diff
@@ -0,0 +1,666 @@
+    value: Option<String>,
+    auto: bool,
+    manual: bool,
+) -> Result<()> {
```
Wrong usage of `--delattr` with argument is ignored and instead interpreted as if argument wasn't present. IMO, this can lead to unintentional deletion of last attribute. The bash script returns an error:
```bash
# mdevctl --version
mdevctl 1.0.0
# mdevctl modify -u 0730d49b-996d-4a51-9075-1afb8db94b57 --addattr a --value b
# mdevctl modify -u 0730d49b-996d-4a51-9075-1afb8db94b57 --addattr a1 --value b1
# mdevctl modify -u 0730d49b-996d-4a51-9075-1afb8db94b57 --delattr=a
# mdevctl list -dvu 0730d49b-996d-4a51-9075-1afb8db94b57
0730d49b-996d-4a51-9075-1afb8db94b57 0.0.26aa vfio_ccw-io manual
  Attrs:
    @{0}: {"a":"b"}

# mdevctl version
0.78
# mdevctl modify -u 48f52ec0-b0fc-42c0-bf58-eb0ff196a780 --addattr=a --value=b
# mdevctl modify -u 48f52ec0-b0fc-42c0-bf58-eb0ff196a780 --addattr=a1 --value=b1
# mdevctl modify -u 48f52ec0-b0fc-42c0-bf58-eb0ff196a780 --delattr=a
mdevctl: option '--delattr' doesn't allow an argument
# mdevctl list -dvu 48f52ec0-b0fc-42c0-bf58-eb0ff196a780
48f52ec0-b0fc-42c0-bf58-eb0ff196a780 0.0.26ab vfio_ccw-io manual
  Attrs:
    @{0}: {"a":"b"}
    @{1}: {"a1":"b1"}
```

jonner replied:
hmm, that's very odd behavior, and I haven't yet figured out how to get the option parsing framework to behave more sanely...

smitterl commented on src/main.rs:
```diff
@@ -0,0 +1,666 @@
+    mdev_type: Option<String>,
+    jsonfile: Option<PathBuf>,
+) -> Result<MDev> {
+    debug!("Starting device '{:?}'", uuid);
```
When trying to start a definition with incorrect attribute, the attribute name is missing from the error message, instead the value is shown:
```bash
# mdevctl --version
mdevctl 1.0.0
# mdevctl list -dvu 0730d49b-996d-4a51-9075-1afb8db94b57
0730d49b-996d-4a51-9075-1afb8db94b57 0.0.26aa vfio_ccw-io manual
  Attrs:
    @{0}: {"a":"b"}
# mdevctl start -u 0730d49b-996d-4a51-9075-1afb8db94b57
Error: Invalid attribute 'b'
```
```bash
# mdevctl version
0.78
# mdevctl list -dvu 48f52ec0-b0fc-42c0-bf58-eb0ff196a780
48f52ec0-b0fc-42c0-bf58-eb0ff196a780 0.0.26ab vfio_ccw-io manual
  Attrs:
    @{0}: {"a":"b"}
# mdevctl start -u 48f52ec0-b0fc-42c0-bf58-eb0ff196a780
Attribute a cannot be set
```

jonner replied:
fixed locally

smitterl commented on src/main.rs:
```diff
@@ -0,0 +1,666 @@
+
+        if dev.is_defined() {
+            return Err(anyhow!(
+                "Device {} on {} already defined, try modify?",
```
IMO, the message should either be shorter `Device {} on {} already defined.` or longer `Device {} on {} already defined. You can modify the existing device instead.". I prefer the first choice. I would want to avoid a question here as it invites the user to answer to the question.

jonner replied:
In this case I just copied the existing error message from mdevctl, but I'm not opposed to shortening it.

jonner commented:
Most of @smitterl's comments are fixed in the new changes above (Thanks!). `make rpm` should now work on RHEL.

cohuck commented on mdevctl.spec.in:
```diff
@@ -5,31 +5,50 @@
+%if 0%{?rhel}
+BuildRequires: rust-toolset
+%else
+%{error: distribution not supported}
```
I'm a bit confused: Above, you also check for Fedora != 34, but here, you seem to fail anything but Fedora 34 and RHEL? Or am I simply not parsing the diff correctly?

jonner replied:
Yes, my implementation is a bit inconsistent / confusing.

The only Fedora distribution that has the necessary dependencies packaged so that we can build the rpm "properly" according to the fedora rust packaging guidelines is Fedora 34. The rpm macros that are used to determine dependencies and build rust crates are provided by the `rust-packaging` package.

In RHEL, there is no `rust-packaging` package, nor are the dependencies packaged. So we have to bundle all of the dependencies (aka "vendor" them). The relevant (but slightly different) rust rpm macros are provided by a `rust-toolset` package.

So the only distributions that I've tested at the moment are Fedora34 and RHEL8/9. It's possible that we could support Fedora33 using a similar vendoring approach, but it would take a little bit of manual work to support it and I wasn't sure if anybody cared. I'm not sure about the state of any other non-Fedora/non-RHEL distributions, so I decided to be conservative and simply error out on anything else.

cohuck replied:
Thanks for the explanation.

Do we have a consensus about which distributions we want to support for `make rpm`? Fedora 34 and later obviously, RHEL to make backporting easy. Is it ok for Fedora 33 to stay on the bash version? Has `make rpm` worked previously on other rpm-based distributions, or is their packaging too different anyway?

jonner replied:
I assume the previous `make rpm` worked on pretty much any version of rhel/fedora because there were so few dependencies. Don't know about other rpm-based distributions though. 

Note that we're only talking about *upstream* rpms here. I view it as a convenience for developers and people that want an easy way to test the git version. Even if we added support for fedora 33 to the upstream spec file, I don't think that the package would be acceptable downstream due to the Fedora rust packaging requirements. So to some extent, I think the decision of whether f33 stays on the bash version is out of our hands.

cohuck replied:
It would be quite nice if it worked on F33 out-of-the-box, but it's probably not a strict requirement.

(FWIW, I played with this a bit on a F33 system, and could not get it to work... not that this says too much :) It does not have the `rust-toolset` package that apparently handles the vendoring on RHEL, and the F34 approach does not work.)

Anyhow, I noticed that you check for == 34 in several places; shouldn't it be >= 34 (unless something changes in later versions?)

jonner replied:
It shouldn't be too hard to support f33 upstream, we could probably just copy the rpm macro out of the rust-toolset package and use that in the upstream spec file. And yes, >= 34 would probably be better.

jonner commented:
OK, I hope this is the final proposal. The only significant change in this last push is improving the rpm a bit per @cohuck's comments. It still doesn't support f33, but we can add that later if necessary (patches welcome ;) )

cohuck commented:
LGTM, I think we can merge this (and do further fixes/improvements on top). @awilliam , what do you think?

awilliam commented:
Thanks all!  /me hands keys to @jonner
